core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -333,46 +340,92 @@ private static void logIfContextNarrowedTimeout(
     context.removeListener(cancellationListener);
-    ScheduledFuture<?> f = deadlineCancellationFuture;
+    ScheduledFuture<?> f = deadlineCancellationSendToServerFuture;
+    if (f != null) {
+      f.cancel(false);
+    }
+
+    f = deadlineCancellationNotifyApplicationFuture;
     if (f != null) {
       f.cancel(false);
     }
   }
 
-  private class DeadlineTimer implements Runnable {
-    private final long remainingNanos;
+  private ScheduledFuture<?> startDeadlineNotifyApplicationTimer(Deadline deadline,
+      final Listener<RespT> observer) {
+    final long remainingNanos = deadline.timeRemaining(TimeUnit.NANOSECONDS);
 
-    DeadlineTimer(long remainingNanos) {
-      this.remainingNanos = remainingNanos;
+    class DeadlineExceededNotifyApplicationTimer implements Runnable {
+      @Override
+      public void run() {
+        Status status = buildDeadlineExceededStatusWithRemainingNanos(remainingNanos);
+        delayedCancelOnDeadlineExceeded(status, observer);
+      }
     }
 
-    @Override
-    public void run() {
-      InsightBuilder insight = new InsightBuilder();
-      stream.appendTimeoutInsight(insight);
-      // DelayedStream.cancel() is safe to call from a thread that is different from where the
-      // stream is created.
-      long seconds = Math.abs(remainingNanos) / TimeUnit.SECONDS.toNanos(1);
-      long nanos = Math.abs(remainingNanos) % TimeUnit.SECONDS.toNanos(1);
-
-      StringBuilder buf = new StringBuilder();
-      buf.append("deadline exceeded after ");
-      if (remainingNanos < 0) {
-        buf.append('-');
+    return deadlineCancellationExecutor.schedule(
+        new LogExceptionRunnable(new DeadlineExceededNotifyApplicationTimer()),
+        remainingNanos,
+        TimeUnit.NANOSECONDS);
+  }
+
+  private Status buildDeadlineExceededStatusWithRemainingNanos(long remainingNanos) {
+    final InsightBuilder insight = new InsightBuilder();
+    stream.appendTimeoutInsight(insight);
+
+    long seconds = Math.abs(remainingNanos) / TimeUnit.SECONDS.toNanos(1);
+    long nanos = Math.abs(remainingNanos) % TimeUnit.SECONDS.toNanos(1);
+
+    StringBuilder buf = new StringBuilder();
+    buf.append("deadline exceeded after ");
+    if (remainingNanos < 0) {
+      buf.append('-');
+    }
+    buf.append(seconds);
+    buf.append(String.format(".%09d", nanos));
+    buf.append("s. ");
+    buf.append(insight);
+
+    return DEADLINE_EXCEEDED.augmentDescription(buf.toString());
+  }
+
+  private void delayedCancelOnDeadlineExceeded(final Status status, Listener<RespT> observer) {
+    class DeadlineExceededSendCancelToServerTimer implements Runnable {
+      @Override
+      public void run() {
+        // DelayedStream.cancel() is safe to call from a thread that is different from where the
+        // stream is created.
+        stream.cancel(status);
+      }
+    }
+
+    deadlineCancellationSendToServerFuture =  deadlineCancellationExecutor.schedule(
```

Comment:
Setting this races with `removeContextListenerAndCancelDeadlineFuture()`. We have two options: 1) let it race. Since we control `DEADLINE_EXPIRATION_CANCEL_DELAY_NANOS` we know the event will fire relatively soon (which is not the case for normal deadlines). 2) Make it thread-safe.

I'm fine with (1), but we should have a comment mentioning the race and that we are okay with it.